src/AutoHashEqualsCached.jl

```diff
@@ -69,135 +88,209 @@ function unpack_name(node::Expr)
         if b.head === :block
             add_fields(field)
         elseif b.head === :const
-            add_field(b.args[1])
+            add_field(__source__, b.args[1])
         elseif b.head === :(::) && b.args[1] isa Symbol
             push!(member_names, b.args[1])
             push!(member_decls, b)
         elseif b.head === :macrocall
-            add_field(b.args[3])
+            add_field(__source__, b.args[3])
         elseif b.head === :function || b.head === :(=) && (b.args[1] isa Expr && b.args[1].head in (:call, :where))
             # :function, :equals:call, :equals:where are defining functions - inner constructors
             # we don't want to permit that if it would interfere with us producing them.
             prevent_inner_constructors &&
-                error("macro @auto_hash_equals_cached should not be used on a struct that declares an inner constructor")
+                error("$(__source__.file):$(__source__.line): macro @auto_hash_equals_cached should not be used on a struct that declares an inner constructor")
         end
     end
-    function add_fields(b::Expr)
+    function add_fields(__source__, b::Expr)
         @assert b.head === :block
         for field in b.args
-            add_field(field)
+            if field isa LineNumberNode
+                __source__ = field
+            else
+                add_field(__source__, field)
+            end
         end
     end
 
     @assert (struct_decl.args[3].head === :block)
-    add_fields(struct_decl.args[3])
+    add_fields(__source__, struct_decl.args[3])
     return (member_names, member_decls)
 end
 
-"""
-    @auto_hash_equals_cached struct Foo ... end
+function check_valid_alt_hash_name(__source__, alt_hash_name)
+    alt_hash_name === nothing || alt_hash_name isa Symbol || Base.is_expr(alt_hash_name, :.) ||
+        error("$(__source__.file):$(__source__.line): invalid alternate hash function name: $alt_hash_name")
+end
 
-Causes the struct to have an additional hidden field named `_cached_hash` that is computed and stored at the time of construction.
-Produces constructors and specializes the behavior of `Base.show` to maintain the illusion that the field does not exist.
-Two different instantiations of a generic type are considered not equal.
+function auto_hash_equals_impl(__source__::LineNumberNode, alt_hash_name, typ::Expr)
+    check_valid_alt_hash_name(__source__, alt_hash_name)
+    struct_decl = get_struct_decl(__source__, typ)
+    @assert struct_decl.head === :struct
 
-Also produces specializations of `Base.hash` and `Base.==`:
+    (type_name, _, _) = unpack_type_name(__source__, struct_decl.args[2])
+    @assert type_name isa Symbol
 
-- `Base.==` is implemented as an elementwise test for `isequal`.
-- `Base.hash` just returns the cached hash value.
-"""
-macro auto_hash_equals_cached(typ::Expr)
-    struct_decl = get_struct_decl(typ)
+    (member_names, _) = get_fields(__source__, struct_decl)
+
+    equalty_impl = foldl(
```

Comment:
Could there be some abstraction here to make this easier to read? This function is a little overwhelming.

Reply:
I added some comments that I think should help.